### DISCUSSION TITLE
Building the distribution images

### DISCUSSION THREAD

**Discussion by bliu13** [SpiralLinux/SpiralLinux-project#228] [Q&A] [answered]
I downloaded the tar.gz files because I was curious on how the distribution is built. I didn't find any documentation for how to set up the environment or how to run the scripts. How can I get started with this?

I don't have any kind of ambitious plans to launch a pointless "competing" distribution, but I want to see what it would take to build the distribution.

**Comment by geckolinux** [comment, marked as answer]
Hi there, sure, it's quite easy and you can even create customized ISOs for your needs.

The tool for building an ISO is maintained by Debian, the package in the main Debian repos is called `live-build` , and it brings several executable scripts and also a system called `live-config` for configuring some aspects of the live ISO during boot. At the bottom of the SpiralLinux website you can find some of the most helpful documentation and examples that I used to get started.

To get started you'll need a Debian 11 system with the `live-build` package installed. I prefer using a VPS or a VM to avoid clutter on my workstation.

At the top level inside the `tar.gz` archives that I upload here you can also find the wrapper script that I created around `live-build` to partially automate some of the initial steps to get the build process going. Basically you just `cd` into the directory and open `spiral` in a text editor and set the `NAME=` variable for the resulting ISO file, and then run `./spiral -i` to start generating it. Other common things that you might want to change first are found at:
- `/config/package-lists/*packages.list.chroot` -- list(s) of Debian packages to install into the live ISO
- `/config/packages/` -- custom DEB packages or DEBs cherry-picked from some other source (In the case of SpiralLinux I add a custom build of Calamares that is only needed for the live ISO, and if needed to fix major bugs and if the package isn't available in Backports I use this method to add newer firmware packages from Testing. This doesn't create a "FrankenDebian" system, because firmware packages are basically for the mini OS that runs on some hardware components and don't really affect the Linux OS itself).
- `/config/includes.chroot/` -- This is where you can overlay files and directories that will be created if they don't exist on the Debian base system, or if they do exist the files placed here will overwrite them. This is one of the keys to creating a unique default configuration for your ISO.
- `/config/hooks/normal/*.hook.chroot` -- These `.hook.chroot` files are BASH scripts that get run inside the `chroot`ed system during the `live-build` process and eventually get wrapped into the final ISO filesystem. This is another key way in addition to `/config/includes.chroot/` to customize the configuration of the final system. Basically any command that you could run on your personal system can be run in the build environment and get included in the final product.

**Reply by fweller** [reply]
@geckolinux, great job!
I have been testing your distro in VM's for the past week.  Overall I am very pleased with your software package choices.  I had some issues with my preferred kwin scripts and the older version of plamashell in debian bullseye, so I updated the repo's to debian testing and the system is working perfectly now.  I am grateful that this distro is 100% married to debian and doesn't try to make it's own repo's.  I was just about to start investigating how to use live-build when I found your instructions here.   Thanks again!

**Reply by geckolinux** [reply]
@fweller Really glad you like it, thanks for saying so!

**Comment by methanoid** [comment]
[edit]

**Comment by manujchandra** [comment]
Hi @geckolinux 

I have 2 questions pertaining to the building of SpiralLinux ISO.

1) In SpiralLinux 11, there is a folder called `SpiralLinux-XFCE_11/config/packages` which contains 3 .deb files. Where are these files getting installed? I was able to find their references in spirallinux-XFCE-packages.list.chroot.

For example, I want to download Veracrypt's .deb files and build an ISO where it's automatically installed. How do I actually install the Veracrypt .deb file from spirallinux-XFCE-packages.list.chroot? What command do I need to put there so the debian file will get installed in the ISO?

2) In `SpiralLinux-XFCE_*/config/hooks/normal/` there is a file called `0999-spirallinux.hook.chroot `
What is calling the hooks? Is my hypothesis correct that all files in the `SpiralLinux-XFCE_*/config/hooks/normal/` folder are automatically executed by lb in numerical order? Is this why its starting with 0999 because you want to execute it at the end?

Kind Regards and thanks for your reply 🙏🏽

**Reply by geckolinux** [reply]
Hi there @manujchandra ,

1. Those were necessary in SpiralLinux 11 because Debian 11 had some very old buggy firmware, so I included those upgraded packages from Debian Testing. Also the Calamares version was very old and limited for Debian 11, so I included a newer version in a DEB, which was only needed for the live system. That would be how you could include Veracrypt in your custom ISO, and you don't need to do anything else apart from placing the DEB file in that directory.
https://live-team.pages.debian.net/live-manual/html/live-manual/customizing-package-installation.en.html#449

2. That's correct. The `live-build` scripts automatically run those hooks during the build process.
https://live-team.pages.debian.net/live-manual/html/live-manual/customizing-contents.en.html#515

**Reply by manujchandra** [reply]
Hello, @geckolinux 

Hope you are well.

The release notes for SpiralLinux 12.* mention that it's using the 6.4 Linux Kernel.

How was this configured in the code? I tried searching for "6.4" in the source, but could not find any references. Please guide. 🙏🏽

Kind Regards.

**Reply by geckolinux** [reply]
Hi there, it comes from the `/etc/apt/preferences.d/99backports` file, which gets obeyed during the `live-build` process just like in a normal installed Debian system.

**Comment by manujchandra** [comment]
Dear @geckolinux 

1) When I update SpiralLinux, will the kernel also get auto updated? Or do I need to manually update them since we are pulling them from stable backports?

2) So if I do not want the latest kernel to be pulled and want to use LTS Kernel, can I remove these lines:

```
Package: linux-image-*
Pin: release a=stable-backports
Pin-Priority: 600

Package: linux-headers-*
Pin: release a=stable-backports
Pin-Priority: 600

```

PS: The intended behavior I want is to use the LTS kernel which ships with Debian 12 and which gets auto updated when the usystem is updated using apt update and upgrade.
and use the following file:

```
# APT pinning preferences - https://wiki.debian.org/AptPreferences

##### Kernel / firmware

Package: firmware-*
Pin: release a=stable-backports
Pin-Priority: 600

##### NVIDIA proprietary drivers

Package: nvidia-driver
Pin: release a=stable-backports
Pin-Priority: 600

```

Thanks 🙏🏽
